[AArch64] Guard against getRegisterBitWidth returning zero in vector instr cost.

davemgreen opened the pull request:
If the getRegisterBitWidth is zero (such as in sme streaming functions), then we could hit a crash from using % RegWidth.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -3248,19 +3248,18 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
     // certain scalar type and a certain vector register width.
-    auto IsExtractLaneEquivalentToZero = [&](const unsigned &Idx,
-                                             const unsigned &EltSz) {
+    auto IsExtractLaneEquivalentToZero = [&](unsigned Idx, unsigned EltSz) {
```
Could you land the NFC changes separately?

davemgreen replied:
Yeah will do.

sdesmalen-arm commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -3277,13 +3276,15 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
         // important.
         UserToExtractIdx[U];
       }
+      if (UserToExtractIdx.empty())
```
I don't see the link with the change in `IsExtractLaneEquivalentToZero`, is this related?

sushgokh commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -3248,19 +3248,18 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
     auto IsUserFMulScalarTy = [](const Value *EEUser) {
       // Check if the user is scalar fmul.
-      const auto *BO = dyn_cast_if_present<BinaryOperator>(EEUser);
+      const auto *BO = dyn_cast<BinaryOperator>(EEUser);
```
EEUser can be null and hence, dont change this

davemgreen replied:
This function is called from users() in two cases. Can you explain why it can be null? Do you have a test case that shows it?

sushgokh replied:
sorry, my bad. You are right. (if_present was required in one of the earlier revisions of the original patch and that lasted till end)

sushgokh commented on llvm/lib/Target/AArch64/AArch64TargetTransformInfo.cpp:
```diff
@@ -3248,19 +3248,18 @@ InstructionCost AArch64TTIImpl::getVectorInstrCostHelper(
           getRegisterBitWidth(TargetTransformInfo::RGK_FixedWidthVector)
               .getFixedValue();
-      return (Idx == 0 || (Idx * EltSz) % RegWidth == 0);
+      return RegWidth != 0 && (Idx == 0 || (Idx * EltSz) % RegWidth == 0);
```
Could you please elaborate why the  `RegWidth` can be 0  in sme streaming functions? If there is some article which I can read about, you can point that.

Can you push the check inside?
`return Idx == 0 || (RegWidth != 0 && (Idx * EltSz) % RegWidth == 0)`

sushgokh reviewed (approved):
LGTM

davemgreen committed:
[AArch64] Guard against getRegisterBitWidth returning zero in vector instr cost.

If the getRegisterBitWidth is zero (such as in sme streaming functions), then
we could hit a crash from using % RegWidth.

It took a while to figure out what was going wrong so there are a few other
minor cleanups here too.

davemgreen commented:
Thanks. The first part was split out into d106a39c3372b924668f203fedbce69aa986cf50.
